### PR TITLE
Fixed issue #3806 by checking if QuantumCircuit args can be cast to int

### LINKED ISSUE #3806: QuantumCircuit constructor fails if n_qubits is np.int64
### What is the expected enhancement?

```
Python 3.5.6 |Anaconda, Inc.| (default, Aug 26 2018, 16:30:03)
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> import qiskit as qk
>>> n = np.int64(12)
>>> qr = qk.QuantumRegister(n)
>>> qc = qk.QuantumCircuit(n)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/qiskit/circuit/quantumcircuit.py", line 157, in __init__
    self.add_register(*regs)
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/qiskit/circuit/quantumcircuit.py", line 541, in add_register
    if register.name in [reg.name for reg in self.qregs + self.cregs]:
AttributeError: 'numpy.int64' object has no attribute 'name'
```

Similar to as was added for `Register`s in #2288, the `QuantumCircuit` constructor should test if a provided argument can be cast to an int before it raises an error.

For reference, here is where `Register` does the check:
https://github.com/Qiskit/qiskit-terra/blob/2ee7a3a/qiskit/circuit/register.py#L39

and where `QuantumCircuit` makes new `Registers` from the provided `int`:
https://github.com/Qiskit/qiskit-terra/blob/703c9a3/qiskit/circuit/quantumcircuit.py#L522

### PR DESCRIPTION
**Pull request by javabster**
### Summary

Fixes #3806 

Added a check to the QuantumCircuit constructor, if an arg is not a classical or quantum register it attempts to cast it to an int before throwing an error

### CODE REVIEW

File: test/python/circuit/test_circuit_properties.py [hunk 1]
```diff
@@ -17,15 +17,30 @@
...
 class TestCircuitProperties(QiskitTestCase):
     """QuantumCircuit properties tests."""
 
+    def test_qarg_numpy_int(self):
+        """Test castable to integer args for QuantumCircuit.
+        """
+        n = np.int64(12)
+        qc1 = QuantumCircuit(n)
+        self.assertEqual(qc1.n_qubits, 12)
+        self.assertEqual(type(qc1), QuantumCircuit)
```
Comment (1ucian0) [inline test/python/circuit/test_circuit_properties.py]: do you mind also adding a test for `int64` in the classical argument?

File: test/python/circuit/test_circuit_properties.py [hunk 2]
```diff
@@ -17,15 +17,30 @@
...
+
+    def test_qarg_numpy_int_exception(self):
+        """Test attempt to pass non-castable arg to QuantumCircuit.
+        """
+        self.assertRaises(CircuitError, QuantumCircuit, 'string')
```
Comment (1ucian0) [inline test/python/circuit/test_circuit_properties.py]: nice test!

File: qiskit/circuit/quantumcircuit.py [hunk 3]
```diff
@@ -123,6 +123,12 @@ class QuantumCircuit:
     extension_lib = "include \"qelib1.inc\";"
 
     def __init__(self, *regs, name=None):
+        if any([not (isinstance(reg, (QuantumRegister, ClassicalRegister)) for reg in regs]):
```
Comment (1ucian0) [inline qiskit/circuit/quantumcircuit.py]: Small syntax error:
```suggestion
        if any([not isinstance(reg, (QuantumRegister, ClassicalRegister)) for reg in regs]):
```

File: test/python/circuit/test_circuit_properties.py (hunk 1 above)
Comment (javabster) [reply test/python/circuit/test_circuit_properties.py]: Sorry, not sure what you mean exactly. Do you mean write a test creating a classical register using int64 and passing the classical register into the quantum circuit?

**Comment by kdk** [comment]
Changes are good to go. To pass the CLA, the email addresses for all the commits in the PR will need to be linked to your github account.

**Comment by javabster** [comment]
> Changes are good to go. To pass the CLA, the email addresses for all the commits in the PR will need to be linked to your github account.

I'm not sure how to fix this, I only used one account to push/pull for this PR. I signed the CLA with that account, I'm not sure why its showing up as 'javabster' and 'Abby Mitchell' when its the same account the whole

**Comment by javabster** [comment]
> > Changes are good to go. To pass the CLA, the email addresses for all the commits in the PR will need to be linked to your github account.
> 
> I'm not sure how to fix this, I only used one account to push/pull for this PR. I signed the CLA with that account and I've added all my emails to the account but it still says its not signed (even after 'rechecking'

Ah never mind it just took a while to refresh after I rechecked. It's good to go now!

File: test/python/circuit/test_circuit_properties.py (hunk 1 above)
Comment (1ucian0) [reply test/python/circuit/test_circuit_properties.py]: Yes, something like `QuantumCircuit(12, np.int64(12))`.

**Pull request merged**